### ISSUE TITLE
Wrong coverage with operators on new lines

### ISSUE THREAD

**Issue by saucisson** [keplerproject/luacov#48]
For instance:

```lua
local s = "some text"
       .. "some other"
```

Writing such code with new lines before operators is sometimes useful to align operator chaining. But `luacov` does not take into account all the lines.

```bash
$ lua -lluacov example.lua
$ luacov
$ cat luacov.report.out
==============================================================================
example.lua
==============================================================================
*0 local s = "some text"
 1        .. "some other"


==============================================================================
Summary
==============================================================================

File        Hits Missed Coverage
--------------------------------
example.lua 1    1      50.00%
--------------------------------
Total       1    1      50.00%
```

The problem also appears in code where the strings are replaced by functions, and the `..` operator is overloaded for function composition, for instance.

**Comment by mpeterv** [comment]
I confirmed that all versions of Lua and LuaJIT don't trigger a debug hook on the first line, so it should be filtered (marked as empty). Unfortunately this is hardly fixable with the current approach, right now filtering works on per line basis and `local s = "some text"` typically shouldn't be considered an empty line. A fix would require writing a full parser emulating parts of Lua compiler that decide where debug hooks can be triggered. In other words, this probably won't be fixed soon sadly.

**Comment by hishamhm** [comment]
On Apr 19, 2016 13:02, "Peter Melnichenko" <notifications@github.com> wrote:
>
> I confirmed that all versions of Lua and LuaJIT don't trigger a debug
hook on the first line, so it should be filtered (marked as empty).
Unfortunately this is hardly fixable with the current approach, right now
filtering works on per line basis and local s = "some text" typically
shouldn't be considered an empty line. A fix would require writing a full
parser emulating parts of Lua compiler that decide where debug hooks can be
triggered. In other words, this probably won't be fixed soon sadly.

In other words, shouldn't this be reported as a Lua issue (a limitation of
Lua's debug hook system) rather than LuaCov?

**Comment by mpeterv** [comment]
@hishamhm Lua gets to attach bytecode to whichever line numbers it wants, but it certainly would be nice if it could export that info. Right now there is "activelines" option for debug.getinfo but it doesn't work for nested functions.

**Comment by mpeterv** [comment]
@hishamhm I forgot to mention here that I implemented a way to get all active lines: https://github.com/mpeterv/deepactivelines. It uses internal Lua C functions but it works, even for LuaJIT. Unfortunately some lines marked as active can't actually trigger the debug hook (this happens when bytecodes for some jumps are unreachable, e.g. `else` after `break`). Still, intersecting sets of lines that current luacov filtering accepts with these "deep" active lines should yield great results.

Unfortunately including or depending on deepactivelines means that installing LuaCov will require compilation. Might as well implement LuaCov core in C to improve performance. @hishamhm what do you think?

**Comment by hishamhm** [comment]
> Lua gets to attach bytecode to whichever line numbers it wants

I understand that this is what it does in practice, but the manual says "the hook is called every time Lua enters a new line of code", and from a Lua user perspective there's no reason why some of the lines of the code that are clearly interpreted as part of the execution don't trigger the line hook. I can see the technical reason why the implementation doesn't trigger the hook, but to me this is a violation of the promise given in the manual. The fact that Lua 5.1 and LuaJIT trigger different lines are also an indication that since the referenece implementation fails to follow the manual specification, other implementors choose to ignore it too. So, I still think this is a wart in Lua.

I think it would make much more sense if the contract of the language was that the hook triggered on each statement, not line.

**Comment by hishamhm** [comment]
@mpeterv I think depending on internal Lua/LuaJIT functions is more problematic than depending on C code. Perhaps making it into an optional dependency for improved accuracy?

**Comment by mpeterv** [comment]
@hishamhm

> I understand that this is what it does in practice, but the manual says "the hook is called every time Lua enters a new line of code", and from a Lua user perspective there's no reason why some of the lines of the code that are clearly interpreted as part of the execution don't trigger the line hook. I can see the technical reason why the implementation doesn't trigger the hook, but to me this is a violation of the promise given in the manual.

It's difficult to precisely define what does "every time Lua enters a new line of code" mean though.

> I think it would make much more sense if the contract of the language was that the hook triggered on each statement, not line.

I agree, it has to be a new hook though. Also, sometimes a statement has no corresponding bytecode (`local a; local b; local c;`), but this optimization can be removed, probably. 

> I think depending on internal Lua/LuaJIT functions is more problematic than depending on C code. Perhaps making it into an optional dependency for improved accuracy?

It works for all versions from Lua 5.1.0 to Lua 5.3.3 and LuaJIT 2.0.0 - 2.1.0-beta2.

**Comment by mpeterv** [comment]
@hishamhm 

> Perhaps making it into an optional dependency for improved accuracy?

Done here: https://github.com/mpeterv/cluacov. Code supporting it is on master branch of luacov. `cluacov` package will depend on luacov so that the whole thing can be installed in one command. It also provides an implementation of debug hook in C which provides 3x-4x speed-up in my tests. I'll make a new luacov release soon and if there will be no issues with `cluacov` it can be merged into luacov one day.

**Comment by mpeterv** [comment]
Released luacov 0.12.0 with cluacov support, it can be installed using `luarocks install cluacov`. @saucisson it seems to handle your snippet correctly.

**Comment by saucisson** [comment]
Thanks!
